downstairs/src/lib.rs

```diff
@@ -1528,22 +1569,27 @@ impl Downstairs {
         Ok(())
     }
 
-    fn is_active(&self, uuid: Uuid) -> bool {
+    fn is_active(&self, connection: UpstairsConnection) -> bool {
         match self.active_upstairs.as_ref() {
             None => false,
             Some((upstairs_connection, _tx)) => {
-                upstairs_connection.upstairs_id == uuid
+                *upstairs_connection == connection
             }
         }
     }
 
-    fn active_upstairs(&self) -> Option<Uuid> {
+    fn active_upstairs(&self) -> Option<UpstairsConnection> {
         self.active_upstairs
             .as_ref()
-            .map(|(upstairs_connection, _tx)| upstairs_connection.upstairs_id)
+            .map(|(upstairs_connection, _tx)| *upstairs_connection)
     }
 
-    async fn clear_active(&mut self) {
+    async fn clear_active(&mut self, upstairs_connection: UpstairsConnection) {
+        // Don't allow another connection to clear us
```

Comment:
How can we get into `clear_active()` without being active?

The two places where I see we call `clear_active()` are both holding the downstairs lock, and both call `ds.is_active()` before calling `clear_active()`.

I know we need some better tests in this area.  I think I was able to find some strange situations by doing both failover testing between different upstairs while restarting downstairs at random.

Reply:
You're right, there isn't a bug here (I posted a followup comment too).